### PR TITLE
[test] Adding health checking integration tests - http

### PR DESCRIPTION
**Pull request by adisuissa**
Signed-off-by: Adi Suissa-Peleg <adip@google.com>

Commit Message: Adding health checking integration tests - http health checks.

Additional Description:
Introducing an integration testing framework for active health checking.
At the moment this includes basic http health check tests, and tcp and gRPC will be added later.
This is the first part to address #12286.

Risk Level: Low (tests only)
Testing: New tests.
Docs Changes: None.
Release Notes: None.

### CODE REVIEW

File: test/integration/health_check_integration_test.cc [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+  }
+
+  // The number of clusters and their names must match the clusters in the CDS integration test
+  // configuration.
+  static constexpr size_t clusters_num_ = 2;
+  std::array<ClusterData, clusters_num_> clusters_{{{"cluster_1"}, {"cluster_2"}}};
+  envoy::type::v3::CodecClientType codec_client_type_{envoy::type::v3::HTTP2};
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, HealthCheckIntegrationTest,
```
Comment (asraa) [inline test/integration/health_check_integration_test.cc:226]: do you expect all or most tests to be parametrized on codec type? if so, maybe use a param for `codec_client_type_`?

Comment (adisuissa) [reply test/integration/health_check_integration_test.cc:226]: The others will be TCP (no codec), and gRPC (only HTTP2).
Only the HTTP tests will be "duplicated" for HTTP/1 and HTTP/2.

Comment (asraa) [reply test/integration/health_check_integration_test.cc:226]: Ah, gotcha! Sounds good, I though maybe this was a base class and TCP and gRPC would be in different classes.

File: test/integration/health_check_integration_test.cc [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    HttpIntegrationTest::initialize();
+
+    // Let Envoy establish its connection to the CDS server.
+    acceptXdsConnection();
+
+    // Expect 1 for the statically specified CDS server.
+    test_server_->waitForGaugeGe("cluster_manager.active_clusters", 1);
+
+    // Register the xds server port in the test framework's downstream listener port map.
+    // test_server_->waitUntilListenersReady();
```
Comment (asraa) [inline test/integration/health_check_integration_test.cc:74]: is this needed?

File: test/integration/health_check_integration_test.cc [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    health_check->mutable_no_traffic_interval()->CopyFrom(
+        Protobuf::util::TimeUtil::MillisecondsToDuration(100));
+    health_check->mutable_unhealthy_threshold()->set_value(1);
+    health_check->mutable_healthy_threshold()->set_value(1);
+    return health_check;
+  }
+
+  // Adds a HTTP active health check specifier to the given cluster with the given timeout,
+  // and waits for the first health check probe to be received.
+  void initHttpHealthCheck(uint32_t cluster_idx, uint32_t timeout = 30) {
```
Comment (asraa) [inline test/integration/health_check_integration_test.cc:132]: is this 30 second timeout a real time? if so is it possible to use simulated time in this class?

**Review by asraa — COMMENTED** [review]
Thanks Adi! The code looks really clean and is very easy to understand (thanks for helping me follow along with the comments on the xDS stream vs the regular endpoints discovered from cds)

File: test/integration/health_check_integration_test.cc (hunk 1 above)
Comment (adisuissa) [reply test/integration/health_check_integration_test.cc]: I've refactored the test suite to use 3 different classes (for HTTP, TCP, and gRPC).
Now there's an `upstream_protocol_` that specifies the required type when creating the host upstream.

File: test/integration/health_check_integration_test.cc (hunk 2 above)
Comment (adisuissa) [reply test/integration/health_check_integration_test.cc]: Removed.

File: test/integration/health_check_integration_test.cc (hunk 3 above)
Comment (adisuissa) [reply test/integration/health_check_integration_test.cc]: Changed to using simulated time.
Thanks for suggesting it.

**Comment by adisuissa** [comment]
@asraa Thanks for your comments. I've refactored the test suite, and changed to simulated time.

**Review by asraa — APPROVED** [review]
Thanks Adi! Clean as always. Looking forward to TCP/gRPC!

**Pull request merged**